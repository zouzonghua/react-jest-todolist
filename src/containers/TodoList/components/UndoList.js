import React, { Component } from 'react';

class UndoList extends Component {
  render() {
    const { list, deleteItem, changeStatus, handleBlur,valueChange } = this.props;
    return (
      <div className="undo-list">
        <div className="undo-list-title">
          正在进行
          <div className="undo-list-count" data-test="count">
            {list.length}
          </div>
        </div>
        <ul className="undo-list-content">
          {list.map((item, index) => (
            <li
              onClick={() => changeStatus(index)}
              className="undo-list-item"
              data-test="list-item"
              key={index}
            >
              {item.status === 'div' ? (
                item.value
              ) : (
                <input
                  className='undo-list-item-input'
                  onBlur={() => handleBlur(index)}
                  onChange={(e) => valueChange(index, e.target.value)}
                  value={item.value}
                  data-test="input" />
              )}
              <span
                className="undo-list-delete"
                onClick={(e) => {
                  e && e.stopPropagation();
                  deleteItem(index)
                }}
                data-test="delete-item"
              >
                -
              </span>
            </li>
          ))}
        </ul>
      </div>
    );
  }
}

export default UndoList;
