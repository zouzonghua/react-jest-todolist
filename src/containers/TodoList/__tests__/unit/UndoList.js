import React from 'react';
import { shallow } from 'enzyme';
import UndoList from '../../components/UndoList';
import { findTestWrapper } from '../../../../utils/testUtils';

describe('UndoList 组件测试', () => {
  it('样式正常', () => {
    const listData = [
      { status: 'div', value: '学习 jest' },
      { status: 'div', value: '学习 TDD' },
    ];
    const wrapper = shallow(<UndoList list={listData} />)
    expect(wrapper).toMatchSnapshot()
  });

  it('未完成列表当数据为空数组 count 数目为0，列表无内容 ', () => {
    const wrapper = shallow(<UndoList list={[]} />);
    const countElem = findTestWrapper(wrapper, 'count');
    const listItems = findTestWrapper(wrapper, 'list-item');
    expect(countElem.text()).toEqual('0');
    expect(listItems.length).toEqual(0);
  });

  it('未完成列表当数据有内容时 count 数目显示数据长度，列表不为空 ', () => {
    const listData = [
      { status: 'div', value: '学习 jest' },
      { status: 'div', value: '学习 TDD' },
    ];
    const wrapper = shallow(<UndoList list={listData} />);
    const countElem = findTestWrapper(wrapper, 'count');
    const listItems = findTestWrapper(wrapper, 'list-item');
    expect(countElem.text()).toEqual('2');
    expect(listItems.length).toEqual(2);
  });

  it('未完成列表当数据有内容时，要存在删除按钮', () => {
    const listData = [
      { status: 'div', value: '学习 jest' },
      { status: 'div', value: '学习 TDD' },
    ];
    const wrapper = shallow(<UndoList list={listData} />);
    const deleteItems = findTestWrapper(wrapper, 'delete-item');
    expect(deleteItems.length).toEqual(2);
  });

  it('未完成列表当数据有内容时，点击某个删除按钮，会调用删除方法 ', () => {
    const listData = [
      { status: 'div', value: '学习 jest' },
      { status: 'div', value: '学习 TDD' },
    ];
    const fn = jest.fn();
    const wrapper = shallow(<UndoList deleteItem={fn} list={listData} />);
    const deleteItems = findTestWrapper(wrapper, 'delete-item');
    deleteItems.at(1).simulate('click');
    expect(fn).toHaveBeenLastCalledWith(1);
  });

  it('当某一项被点击时，触发执行 changeStatus 函数 ', () => {
    const listData = [
      { status: 'div', value: '学习 jest' },
      { status: 'div', value: '学习 TDD' },
    ];
    const fn = jest.fn();
    const index = 1;
    const wrapper = shallow(<UndoList changeStatus={fn} list={listData} />);
    const listItems = findTestWrapper(wrapper, 'list-item');
    listItems.at(1).simulate('click');
    expect(fn).toHaveBeenLastCalledWith(index);
  });

  it('当某一项 status 是 input 时，展示输入框', () => {
    const listData = [
      { status: 'div', value: '学习 jest' },
      { status: 'input', value: '学习 TDD' },
    ];
    const wrapper = shallow(<UndoList list={listData} />);
    const inputItems = findTestWrapper(wrapper, 'input');
    expect(inputItems.length).toBe(1);
  });

  it('当某一个输入框失去焦点时，触发执行 handleBlur 方法 ', () => {
    const listData = [
      { status: 'div', value: '学习 TDD' },
      { status: 'input', value: '学习 jest' },
    ];
    const fn = jest.fn();
    const wrapper = shallow(<UndoList handleBlur={fn} list={listData} />);
    const inputItems = findTestWrapper(wrapper, 'input');
    inputItems.simulate('blur');
    expect(fn).toHaveBeenLastCalledWith(1);
  });

  it('当某一个输入框变更时，触发执行 valueChange 方法 ', () => {
    const listData = [
      { status: 'div', value: '学习 TDD' },
      { status: 'input', value: '学习 jest' },
    ];
    const fn = jest.fn();
    const value = '学习 React';
    const wrapper = shallow(<UndoList valueChange={fn} list={listData} />);
    const inputItems = findTestWrapper(wrapper, 'input');
    inputItems.simulate('change', {
      target: { value },
    });
    expect(fn).toHaveBeenLastCalledWith(1, value);
  });
});
