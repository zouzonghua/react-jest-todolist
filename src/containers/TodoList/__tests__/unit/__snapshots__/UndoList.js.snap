// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UndoList 组件测试 样式正常 1`] = `
<div
  className="undo-list"
>
  <div
    className="undo-list-title"
  >
    正在进行
    <div
      className="undo-list-count"
      data-test="count"
    >
      2
    </div>
  </div>
  <ul
    className="undo-list-content"
  >
    <li
      className="undo-list-item"
      data-test="list-item"
      key="0"
      onClick={[Function]}
    >
      学习 jest
      <span
        className="undo-list-delete"
        data-test="delete-item"
        onClick={[Function]}
      >
        -
      </span>
    </li>
    <li
      className="undo-list-item"
      data-test="list-item"
      key="1"
      onClick={[Function]}
    >
      学习 TDD
      <span
        className="undo-list-delete"
        data-test="delete-item"
        onClick={[Function]}
      >
        -
      </span>
    </li>
  </ul>
</div>
`;
